gradually-update/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+			// If current_value is zero, it's length is zero, so can't compare with target_value.len().
+			ensure!(update.per_block.len() == update.target_value.len() && update.per_block.len() <= 16, Error::<T>::InvalidTargetValue);
+
+			let mut gradually_updates = GraduallyUpdates::get();
+			ensure!(!gradually_updates.contains(&update), Error::<T>::GraduallyUpdateHasExisted);
+
+			gradually_updates.push(update.clone());
+			GraduallyUpdates::put(gradually_updates);
+
+			Self::deposit_event(RawEvent::GraduallyUpdate(update.key, update.per_block, update.target_value));
+		}
+
+		/// Cancel gradually_update to adjust numeric parameter. This is a root call.
+		pub fn cancel_gradually_update(origin, key: StorageKey) {
+			ensure_root(origin)?;
+
+			let gradually_updates: Vec<GraduallyUpdate> = GraduallyUpdates::get()
+				.into_iter()
+				.filter(|item| item.key != key)
+				.collect();
+
+			ensure!(GraduallyUpdates::get().len() - gradually_updates.len() == 1, Error::<T>::CancelGradullyUpdateNotExisted);
+			GraduallyUpdates::put(gradually_updates);
+
+			Self::deposit_event(RawEvent::CancelGraduallyUpdate(key));
+		}
+
+		/// Update gradually_update to adjust numeric parameter.
+		fn on_finalize(now: T::BlockNumber) {
+			Self::_on_finalize(now);
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	fn _on_finalize(now: T::BlockNumber) {
+		if now < GraduallyUpdateBlockNumber::<T>::get() + T::UpdateFrequency::get() {
+			return;
+		}
+
+		let mut gradually_updates = GraduallyUpdates::get();
+		for (i, update) in gradually_updates.clone().iter().enumerate() {
+			let current_value = storage::unhashed::get::<StorageValue>(&update.key).unwrap_or_default();
+			let current_value_u128 = u128::from_le_bytes(Self::convert_vec_to_u8(&current_value));
+
+			let frequency_u128 =
+				u128::from_le_bytes(Self::convert_vec_to_u8(&T::UpdateFrequency::get().encode().to_vec()));
+			let step = u128::from_le_bytes(Self::convert_vec_to_u8(&update.per_block));
+			let step_u128 = step.checked_mul(frequency_u128).unwrap();
+
+			let target_u128 = u128::from_le_bytes(Self::convert_vec_to_u8(&update.target_value));
+
+			let new_value_u128: u128;
+			if current_value_u128 > target_u128 {
+				new_value_u128 = (current_value_u128.checked_sub(step_u128).unwrap()).max(target_u128);
+			} else {
+				new_value_u128 = (current_value_u128.checked_add(step_u128).unwrap()).min(target_u128);
+			}
+
+			// current_value equal target_value, remove gradually_update
+			if new_value_u128 == target_u128 {
+				gradually_updates.remove(i);
+			}
+
+			let mut value = new_value_u128.encode();
+			value.truncate(update.target_value.len());
+
+			storage::unhashed::put(&update.key, &value);
+
+			Self::deposit_event(RawEvent::GraduallyUpdateBlockNumber(now, update.key.clone(), value));
+		}
+
+		// gradually_update has finished. Remove it from GraduallyUpdates.
+		if gradually_updates.len() < GraduallyUpdates::get().len() {
```

Comment:
```suggestion
		if gradually_updates.len() < GraduallyUpdates::decode_len() {
```

decode_len read the vec length without decode the whole vec